signer/src/block_notifier.rs

```diff
@@ -0,0 +1,150 @@
+                _ = ping_interval.tick() => {
+                    if client.ping().is_err() {
+                        error!("Ping failed, attempting to reconnect...");
+                        retry_attempts += 1;
+                        if retry_attempts >= max_retry_attempts {
+                            error!("Max retry attempts reached, giving up.");
+                            continue_loop = false;
+                        }
+                    } else {
+                        retry_attempts = 0; // Reset retry attempts on successful ping
+                    }
+                }
+                // Periodic block header subscription
+                _ = subscribe_interval.tick() => {
+                    match client.block_headers_subscribe() {
+                        Ok(HeaderNotification { header, .. }) => {
+                            let block_hash = header.block_hash();
+
+                            if sender.send(Ok(block_hash)).await.is_err() {
+                                warn!("Receiver dropped, stopping notify loop.");
+                                continue_loop = false;
+                            }
+                        }
+                        Err(e) => {
+                            error!("Error subscribing to block headers: {}", e);
+                            retry_attempts += 1;
+                            if retry_attempts >= max_retry_attempts {
+                                error!("Max retry attempts reached, giving up.");
+                                continue_loop = false;
+                            }
+                        }
+                    }
+                }
+            }
+            sleep(retry_interval).await;
+        }
+    }
+
+    /// Starts the notification loop in a new task.
+    ///
+    /// # Arguments
+    ///
+    /// * `sender` - The sender to which block headers will be sent.
+    fn start_notifying(&self, sender: Sender<Result<BlockHash, ElectrumError>>) {
+        let client = Arc::clone(&self.client);
+        let retry_interval = self.retry_interval;
+        let max_retry_attempts = self.max_retry_attempts;
+        let ping_interval = self.ping_interval;
+        let subscribe_interval = self.subscribe_interval;
+
+        task::spawn(Self::notify_loop(
+            client,
+            retry_interval,
+            max_retry_attempts,
+            ping_interval,
+            subscribe_interval,
+            sender,
+        ));
+    }
+}
+
+impl BlockNotifier for ElectrumBlockNotifier {
+    type Error = ElectrumError;
+
+    /// Subscribes to a stream of block headers.
+    ///
+    /// # Returns
+    ///
+    /// A pinned boxed stream of block headers.
+    fn subscribe(&self) -> Pin<Box<dyn Stream<Item = Result<BlockHash, Self::Error>> + Send>> {
+        let (sender, receiver) = channel(100); // Bounded channel to control memory usage
+        self.start_notifying(sender);
+        Box::pin(ReceiverStream::new(receiver))
```

Comment:
If you update the trait to use generics, we can return the receiver stream without boxing here
```suggestion
        ReceiverStream::new(receiver)
```